hivemq-edge/src/main/java/com/hivemq/edge/modules/adapters/impl/polling/PollingTask.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2019-present HiveMQ GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hivemq.edge.modules.adapters.impl.polling;
+
+import com.hivemq.adapter.sdk.api.events.EventService;
+import com.hivemq.adapter.sdk.api.events.model.Event;
+import com.hivemq.configuration.service.InternalConfigurations;
+import com.hivemq.edge.modules.api.adapters.ProtocolAdapterPollingSampler;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.util.ExceptionUtils;
+import org.jetbrains.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class PollingTask implements Runnable {
+
+    private static final Logger log = LoggerFactory.getLogger(PollingTask.class);
+
+    private final @NotNull ProtocolAdapterPollingSampler sampler;
+    private final @NotNull ScheduledExecutorService scheduledExecutorService;
+    private final @NotNull EventService eventService;
+    private final @NotNull AtomicInteger watchdogErrorCount = new AtomicInteger();
+    private final @NotNull AtomicInteger applicationErrorCount = new AtomicInteger();
+
+    private volatile long nanosOfLastPolling;
+    private final @NotNull AtomicBoolean continueScheduling = new AtomicBoolean(true);
+
+
+    public PollingTask(
+            final @NotNull ProtocolAdapterPollingSampler sampler,
+            final @NotNull ScheduledExecutorService scheduledExecutorService,
+            final @NotNull EventService eventService) {
+        this.sampler = sampler;
+        this.scheduledExecutorService = scheduledExecutorService;
+        this.eventService = eventService;
+    }
+
+    @Override
+    public void run() {
+        try {
+            nanosOfLastPolling = System.nanoTime();
+
+            // synchronized with stop so that the change of the executor future field and the setting of the continueScheduling
+            // has exclusive access by either of the two methods.
+            synchronized (this) {
```

Comment:
Redundant/obsolete?

Reply:
yeah I wanted to return/expose the future of the sampler.execute in stop(), but it changes frequently and is often null after the execution and before the next execution, so this seemed useless to me as stop basically kills the next poll attempt anyway, which seems semantically good enough.